File: src/main/java/answer/chapter5/Problem5_1.java
package answer.chapter5;

import jm.app.algebra.Matrix;
import jm.app.optimize.Optimizer;
import jm.app.optimize.SteepestDescentOptimizer;
import jm.app.optimize.TargetFunction;

import java.math.BigDecimal;

/**
 * 利用SteepDescentOptimizer，分别计算出目标函数
 * F(X)=4*x1^3+2*x2^2+3*x3+x4在x1=7，x2=5，x3=-1，x4=16时
 * 关于x1，x2，x3，x4的偏导数。
 *
 * 思路：将F(X;A)转换为关于A的函数F(A;X)=x1^3*a1+x2^2*a2+x3*a3+x4*a4之后再计算,
 * 其中a1=4, a2=2, a3=3, a4=1，此时params是X, args是A
 */
public class Problem5_1 {

    static class MyTargetFunction implements TargetFunction {
        @Override
        public BigDecimal fun(Matrix params, Matrix args) {
            BigDecimal x1 = params.getValue(0, 0);
            BigDecimal x2 = params.getValue(0, 1);
            BigDecimal x3 = params.getValue(0, 2);
            BigDecimal x4 = params.getValue(0, 3);
            BigDecimal a1 = args.getValue(0, 0);
            BigDecimal a2 = args.getValue(0, 1);
            BigDecimal a3 = args.getValue(0, 2);
            BigDecimal a4 = args.getValue(0, 3);
            BigDecimal result = x1.pow(3).multiply(a1).add(x2.pow(2).multiply(a2)).add(x3.multiply(a3)).add(x4.multiply(a4));
            return result;
        }
    }

    public static void main(String[] args) {
        SteepestDescentOptimizer optimizer = new SteepestDescentOptimizer();
        Matrix params = new Matrix(1, 4);
        params.setValue(0, 0, 7);
        params.setValue(0, 1,  5);
        params.setValue(0, 2, -1);
        params.setValue(0, 3, 16);
        Matrix arguments = new Matrix(1, 4);
        arguments.setValue(0, 0, 4);
        arguments.setValue(0, 1, 2);
        arguments.setValue(0, 2, 3);
        arguments.setValue(0, 3, 1);
        MyTargetFunction myTargetFunction = new MyTargetFunction();
        BigDecimal partialDerivativeOfX1 = optimizer.calcPartialDerivative(myTargetFunction, params, 0, arguments);
        BigDecimal partialDerivativeOfX2 = optimizer.calcPartialDerivative(myTargetFunction, params, 1, arguments);
        BigDecimal partialDerivativeOfX3 = optimizer.calcPartialDerivative(myTargetFunction, params, 2, arguments);
        BigDecimal partialDerivativeOfX4 = optimizer.calcPartialDerivative(myTargetFunction, params, 3, arguments);
        System.out.println(partialDerivativeOfX1);
        System.out.println(partialDerivativeOfX2);
        System.out.println(partialDerivativeOfX3);
        System.out.println(partialDerivativeOfX4);
    }

}

File: py/answer/chapter5/problem5_1.py
import py.optimize.target_function as tf
import py.optimize.steepest_descent_optimizer as sdo
import numpy as np


class TestTargetFunction(tf.TargetFunction):

    def fun(self, params, args):
        x1 = params[0]
        x2 = params[1]
        x3 = params[2]
        x4 = params[3]
        a1 = args[0]
        a2 = args[1]
        a3 = args[2]
        a4 = args[3]
        res = a1 * np.power(x1, 3) + a2 * np.power(x2, 2) + a3 * x3 + a4 * x4
        return res

def test_calc_partial_derivative():
    params = np.array([7.0, 5.0, -1.0, 16.0])
    args = np.array([4.0, 2.0, 3.0, 1.0])
    target_function = TestTargetFunction()
    optimizer = sdo.SteepestDescentOptimizer(epoch_num = 10000)
    partial_derivative_of_x1 = optimizer.calc_partial_derivative(target_function, params, 0, args)
    partial_derivative_of_x2 = optimizer.calc_partial_derivative(target_function, params, 1, args)
    partial_derivative_of_x3 = optimizer.calc_partial_derivative(target_function, params, 2, args)
    partial_derivative_of_x4 = optimizer.calc_partial_derivative(target_function, params, 3, args)
    print(partial_derivative_of_x1)
    print(partial_derivative_of_x2)
    print(partial_derivative_of_x3)
    print(partial_derivative_of_x4)

if __name__ == "__main__":
    test_calc_partial_derivative()